ImprovedOpAmpLimited

AHaumer opened the pull request:
This PR replaces #4485 which will be closed but we still can look up the discussions.
The following explanation could be used for future release notes:
Modelica.Electrical.Analog.Ideal.IdealizedOpAmpLimited has been declared as deprecated since some of the results (Modelica.Electrical.Analog.Examples.OpAmps) could not be compared between different tools, because the usage of noEvent() and smooth() for saturation prevented exact events in switching applications. 
As a replacement Modelica.Electrical.Analog.Ideal.ImprovedOpAmpLimited has been implemented, thoroughly tested with different tools and used in the above mentioned examples. Saturation has been designed simpler, events are triggered by two Boolean variables that indicate saturation. Additionally a firstOrder models rise time of the OpAmp. Initialization is now without homopotopy acting on the result of the firstOrder (before saturation acts on the output).

HansOlsson commented:
I'll look into it.
Just a quick style-question - is there a particular reason you switched from `SI.Frequency` (etc) to `Modelica.Units.SI.Frequency`?

AHaumer commented:
@HansOlsson this was by accident. 
I copied all the OpAmp stuff out of MSL, then developed and tested and copied back in. 
You're right, looks clumsy. Therefore replaced all `Modelica.Units.SI.` by `SI.`

maltelenz commented:
Without looking at the actual changes, I ran the full MSL tests, and got result differences in `Adder`, `ControlCircuit` and `Subtracter`.

Blue is the reference data, orange is the new result with this PR.

`Adder`:
<img width="960" alt="image" src="https://github.com/user-attachments/assets/5ac79e30-183a-4004-8f38-541df44b3362" />

In `ControlCircuit` it looks like an event became a softer transition at `0.5`? This is zoomed in:

<img width="960" alt="image" src="https://github.com/user-attachments/assets/96b91993-c43c-4346-a130-9fff2d6b1de0" />

`Subtracter`:
<img width="960" alt="image" src="https://github.com/user-attachments/assets/a81106de-0207-4665-8dce-880fb0a9e8b3" />

AHaumer commented:
After merging this PR, I'll prepare another PR containing a remake of Modelica.Electrical.Examples.ChuaCircuit using realistic parameters (mH and pF insteaf of H and F), showing the path from 2 periodic attractors to chaos and an additonal example demonstrating how the characteristic of Chua's diode could be implemented with 2 OpAmp-circuits called NIC (negative impedance converter) - that example relies on the OpAmp-model of this PR.

HansOlsson commented on Modelica/Electrical/Analog/Examples/OpAmps/package.order:
```diff
@@ -12,7 +12,7 @@ VoltageFollower
 Comparator
 InvertingSchmittTrigger
 SchmittTrigger
+LCOscillator
```
Any reason for the changed order?

AHaumer replied:
Yes I tried to order the examples according to the problems they did rise during the last regression tests ;-)
I see no big harm in changing the order ...

HansOlsson commented on Modelica/Electrical/Analog/Examples/package.order:
```diff
@@ -31,6 +31,6 @@ ParallelResonance
 DemoPowerSupplyWithBuffer
-Lines
 OpAmps
+Lines
```
Also question about the order.

AHaumer replied:
Yes I tried to order the examples according to the problems they did rise during the last regression tests ;-)
I see no big harm in changing the order ...

HansOlsson commented on Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo:
```diff
@@ -0,0 +1,110 @@
+    vps = Vps;
+    vns = Vns;
+  end if;
+  assert(vps > vns, "Supply voltages are not consistent: vps <= vns");
```
There's no longer any need to repeat the condition; see https://github.com/modelica/ModelicaSpecification/pull/3641

AHaumer replied:
Sorry I missed that discussion. Accepted.

HansOlsson commented on Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo:
```diff
@@ -0,0 +1,110 @@
+  in_p.i = 0;
+  in_n.i = 0;
+  //firstOrder to model Rise time
+  der(v_int) = (V0*v_in - v_int)/(Tr/log(9));
```
I don't fully get the log(9), and I think it should be explained/replaced by some final parameter.
I assume it is related to 10-90% rise time.

I also wonder if it would be possible to have an option to de-activate the firstOrder part.

AHaumer replied:
Yes we could use a choice between fristOrder and simple gain - I'll try that,
Maybe dependent on the size of the time constant (too large is bad, too small is bad).
And I'll name the time constant differently and remove the log(9), this is not exact.
The log(9) stems from:
0.1 = (1 - exp(-t10/Tau)
0.9 = (1 - exp(-t90/Tau)
Tr = t90 - t10 = Tau*ln(9)

AHaumer replied:
@HansOlsson I think it is a good idea to have an option to deavtivate the firstOrder part. 
On the other hand this makes initialization much more complicated. 
With firstOrder it's easy: Just initialize the output of the firstOrder.
With simple gain I'd have to fiddle around with homotopy (to be able to fix the output at positive or negative limit) which isn't that easy when limits are not constant (parameters).

HansOlsson commented on Modelica/Electrical/Analog/Examples/OpAmps/Differentiator.mo:
```diff
@@ -42,7 +41,6 @@ equation
     annotation (Line(points={{10,-10},{40,-10}}, color={0,0,255}));
   annotation (Documentation(info="<html>
 <p>This is a (inverting) differentiating amplifier. Resistance R can be chosen, capacitance C is defined by the desired time constant resp. frequency.</p>
-<p>Note: <code>vOut</code> measure the negative output voltage.</p>
```
I noticed this comment has been erased for all of the examples, even though the component remains.
I assume it is intentional and because it is self-explanatory, right?

AHaumer replied:
Yes I changed the polarity of measurement of the output voltage, therefore removed the line in the documentation. 
I saw that it is much easier to understand: This is an inverting circuit, and I see this fact looking at the output voltage without thinking about the polarity of the voltage sensor.

HansOlsson reviewed (commented):
In general good, but I would need some feedback on the filter-part.

AHaumer commented:
> In general good, but I would need some feedback on the filter-part.

You mean the firstOrder? Give me some time over the weekend to design initialization.

AHaumer commented:
@HansOlsson I hope I addressed all your concerns, please re-review.
Now the firstOrder is optional (and therefore the initialization for it). It is needed only in 2 examples: MultiVibrator and SignalGenerator. The results seem to be the same in Dymola, OpenModelica and SystemModeler. 
@casella @maltelenz @henrikt-ma may I kindly ask for your review?

AHaumer commented:
Ok there a some cosmetic changes (my fault I know) in the examples.
The main work is a nearly complete redesign of the OpAmp model, some more explanations:

- The differential input voltage is either just amplified by V0 or (optional) additionally delayed with a firstOrder, this gives the fictional intermediate voltage
-  If the optional fristOrder is activated this intermediate voltage can be intialized as Linear, UpperLimit, LowerLimit.
- This intermediate voltage gets limited between positive and negative supply without triggering events to get output voltage.
- Events are triggered by 2 Boolean indicators of saturation, so saturation should be documented exactly.

Should I include a test model (meaybe in ModelicaTest?) measuring the rise time?
To meet the RiseTime stated in data sheet (e.g. LM741 0.3 µs), the time constant has to be set to V0/log(9)*RiseTime. 
RiseTime is measured in a voltageFollower circuit and defined by `Tr = t90 - t10` where 
`0.1 = 1 - exp(-t10/Tau)` and `0.9 = 1 - exp(-t90/Tau)`, using a step input. The differential equation for the voltageFollower is: 
`Tau*(der(vOut)) = (V0*(vStep - vOut)) - vOut` taking the firstOrder into account.
Many applications are nearly not affected by this firstOrder, so it is deactivated. 
For some applications the user would have to carefully compare time constant and transient behaviour of his input. 
For some applications (MultiVibrator and SignalGenerator) the firstOrder is essential to eases numerical solution.

AHaumer commented:
wait wait wait ... silly me: I forgot to check / adapt the comparisonSignals.

Then I want another change of the order of the examples. After that we get:

- 10 examples using the ready-to-use models from OpAmpCircuits
- 4 non-autonomous circuits and 1 autonomous (up to now uncritical) switching circuit
- 2 autonomous critical circuits

And I'll implement the correct measurement of rise time in ModelicaTest.

All done after the following commit (lastest this evening).

christiankral commented on Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo:
```diff
@@ -0,0 +1,128 @@
+<p>Supply voltage is either defined by parameter <code>Vps</code> and <code>Vns</code> or by (optional) pins <code>s_p</code> and <code>s_n</code>.<br>
+In the first case the necessary power is drawn from an implicit internal supply, in the second case from the external supply.</p>
+<p>
+For most applications it is sufficient ot use default settings <code>Advanced.useFirstOrder=false</code>. 
```
```suggestion
For most applications it is sufficient ot use default settings <code>useFirstOrder=false</code>. 
```

Using `Advanced.` somehow suggests that `Advanced` is a Modelica class.

christiankral commented on Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo:
```diff
@@ -0,0 +1,128 @@
+<p>
+For most applications it is sufficient ot use default settings <code>Advanced.useFirstOrder=false</code>. 
+In this case the intermediate voltage <code>v_int</code> is simply <code>V0*v_in</code>.<br>
+In some applications it might be necessary to set <code>Advanced.useFirstOrder=true</code> 
```
```suggestion
In some applications it might be necessary to set <code>useFirstOrder=true</code> 
```

christiankral commented on Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo:
```diff
@@ -0,0 +1,128 @@
+it is unlikely that it is necessary to change the start values and / or to declare them as fixed.
+</p>
+<p>
+If it is necessary to use <code>Advanced.useFirstOrder=true</code>, it is essential to initialize intermediate voltage <code>v_int</code> correctly.<br>
```
```suggestion
If it is necessary to use <code>useFirstOrder=true</code>, it is essential to initialize the intermediate 
```

christiankral commented on Modelica/Electrical/Analog/Types/InitOpAmp.mo:
```diff
@@ -0,0 +1,6 @@
+    Linear "Initialization of firstOrder linearly dependent on input",
+    UpperLimit "Initialization of firstOrder at positive supply",
+    LowerLimit "Initialization of firstOrder at negative supply")
+  "Enumeration defining initialization of the IdealizedOpAmpLimited" annotation (Evaluate=true);
```
```suggestion
  "Enumeration defining initialization of the ImprovedOpAmpLimited" annotation (Evaluate=true);
```

The type `InitOpAmp` is not used in `IdealizedOpAmpLimited`

christiankral reviewed (commented):
I wonder if `ImprovedOpAmpLimited` is a good choice for a component name. One day `IdealizedOpAmpLimited` will be removed and we might not remember exactly what was improved ...

christiankral commented:
> Sorry @maltelenz I forgot to mention:
> 
> * In some examples I flipped the output voltage sensor vertically. That explains the drastic differences.

We might want to add a comment in `Modelica.Electrical.Analog.UsersGuide.ReleaseNotes` in case someone is wondering why the results appear differently in MSL 4.1.0.

AHaumer commented:
@christiankral I accepted all your valuable suggestions with [9209342](https://github.com/modelica/ModelicaStandardLibrary/pull/4643/commits/92093421f165eb2446075d7ed9647ee8869586fb).
Regarding the naming, in Modelica.Electrical.Analog.Ideal we have;

- IdealOpAmp (infinite gain, no saturation)
- IdealOpAmp3Pin (infinite gain, no saturation)
- IdealOpAmpLimited (infinite gain, saturation, compulsory supply)
- deprecated IdealizedOpAmpLimited (finite gain, saturation, optional supply)
- new ImprovedOpAmpLimited (finite gain, saturation, optional supply, optional firstOrder rise of output, stability)

The new one is not only improved comparing wih the deprecated one, but also with the older three.

christiankral reviewed (approved):
OK then, so we stick with the name `ImprovedOpAmpLimited`.

If I everybody else agrees with the name, it's fine with me, too...

AHaumer commented:
Text for release notes:
ImprovedOpAmpLimited replaces IdealizedOpAmpLimited (which is deprecated), with finite gain, saturation, optional supply and optional firstOrder rise of output which improves stability.
The examples in OpAmps are adapted to use the new ImprovedOpAmpLimited and harmonized polarity of output voltage sensor.
@GallLeo @MatthiasBSchaefer please provide new reference results for all examples in Modelica.Electrical.Analog.Examples.OpAmps after this PR has been merged.

casella commented:
Unfortunately the discussion on the naming was too quick for me to react. I don't really think that "Improved" is a good attribute for any Modelica Standard Library model. Same as "New". It is new or improved now, and then it will be soon obsolete, as all things human.

The name should describe **what the model represents**, not **how its development history went**.

Why don't we just call it OpAmpLimited? It's not idealized, because it has a finite gain. So, it's just an op amp. @AHaumer do you agree?
